File: ios/Avatar/Avatar/Models/RobotState.swift
//
//  RobotState.swift
//  Avatar
//
//  Core data model for the stick-figure companion.
//  Ported from Android: RobotState.kt
//

import Foundation

/// Top-level behavior state (FSM)
enum RobotMode: Equatable {
    /// Idle, waiting for interaction. Stick figure does random goofy antics.
    case idle

    /// User tapped or said wake word — waiting for speech input.
    case listening

    /// Stick figure is speaking (TTS active). Mouth + arm gestures animate.
    case speaking

    /// Processing request (e.g., waiting on LLM response).
    case thinking

    /// Camera is active (rear camera, on-demand). Stick figure holds a "camera" pose.
    case looking
}

/// Emotional state — drives expression and tone
enum Emotion: CaseIterable {
    case neutral
    case happy
    case curious
    case surprised
    case shy
    case sleepy
    case sad
    /// Goofy / silly mood — for random antics
    case goofy

    var intensity: Float {
        switch self {
        case .neutral:    return 0.5
        case .happy:      return 0.8
        case .curious:    return 0.6
        case .surprised:  return 0.9
        case .shy:        return 0.4
        case .sleepy:     return 0.3
        case .sad:        return 0.3
        case .goofy:      return 0.9
        }
    }
}

/// Current stick figure state, consumed by UI and behavior engine.
struct RobotState {
    var mode: RobotMode = .idle

    /// Current emotion
    var emotion: Emotion = .neutral

    /// Whether TTS is currently producing audio
    var isSpeaking: Bool = false

    /// Last recognized user utterance
    var lastUserText: String? = nil

    /// Response text to speak
    var responseText: String? = nil

    /// Blink trigger — UI toggles each time this changes
    var blinkTrigger: Int = 0

    /// Random antic trigger — increments to make the stick figure do something goofy
    var anticTrigger: Int = 0

    /// When LOOKING: a description of what the camera saw (filled by LLM or local logic)
    var cameraObservation: String? = nil
}

File: android/app/src/main/java/com/rd/avatar/robot/RobotState.kt
package com.rd.avatar.robot

/**
 * Core data model for the robot companion.
 */

/** Top-level behavior state (FSM) */
enum class RobotMode {
    /** Idle, waiting for interaction. Eyes wander, occasional blinks. */
    IDLE,

    /** User tapped or said wake word — waiting for speech input. */
    LISTENING,

    /** Robot is speaking (TTS active). Mouth animates. */
    SPEAKING,

    /** Human face detected — engaged, eyes follow face. */
    WATCHING,

    /** Processing request (e.g. waiting on LLM response). */
    THINKING
}

/** Emotional state — drives expression and tone */
enum class Emotion(val intensity: Float) {
    NEUTRAL(0.5f),
    HAPPY(0.8f),
    CURIOUS(0.6f),
    SURPRISED(0.9f),
    SHY(0.4f),
    SLEEPY(0.3f),
    SAD(0.3f)
}

/**
 * Current robot state, consumed by UI and behavior engine.
 */
@androidx.compose.runtime.Stable
data class RobotState(
    val mode: RobotMode = RobotMode.IDLE,

    /** Detected face bounding box center, normalized 0..1, or null */
    val faceTargetX: Float? = null,
    val faceTargetY: Float? = null,

    /** Current emotion */
    val emotion: Emotion = Emotion.NEUTRAL,

    /** Whether TTS is currently producing audio */
    val isSpeaking: Boolean = false,

    /** Last recognized user utterance */
    val lastUserText: String? = null,

    /** Response text to speak */
    val responseText: String? = null,

    /** Blink trigger — UI toggles each time this changes */
    val blinkTrigger: Long = 0L,

    /** Time since last face detection in ms */
    val msSinceLastFace: Long = 0L
)
